REGRESSION (275943@main): Apps using Aspects framework can crash after ApplicationStateTracker adds KVObserver

jernoble opened the pull request:
#### 759b0ea6d2cd7e5196e5d2a78931d8a71c144a88
<pre>
REGRESSION (275943@main): Apps using Aspects framework can crash after ApplicationStateTracker adds KVObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=272532">https://bugs.webkit.org/show_bug.cgi?id=272532</a>
<a href="https://rdar.apple.com/125932015">rdar://125932015</a>

Reviewed by Brent Fulgham.

Aspects.framework will crash if it tries to add a &quot;hook&quot; for KVO after an observer has already
been added to the target object. Add a linked-on-or-after check which disables this KVO, giving
apps time to move off of using Aspects in their production code.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(-[WKUIWindowSceneObserver setObservedWindow:]):

Canonical link: <a href="https://commits.webkit.org/277396@main">https://commits.webkit.org/277396@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ac2c3eb451cca4f7a986fe69dd83c06721c81d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50148 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43513 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38641 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24254 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40908 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19963 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21701 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42077 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5508 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/40743 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43810 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52026 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46955 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45944 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23771 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44982 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10478 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24561 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54453 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23489 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11170 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -37,9 +37,11 @@
 #import <wtf/spi/cocoa/SecuritySPI.h>
 #import <wtf/text/TextStream.h>
 
+
```
Do we want this extra whitespace?

jernoble replied:
Nope!

brentfulgham commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -78,6 +80,9 @@ - (id)initWithParent:(WebKit::ApplicationStateTracker&)parent
 
 - (void)setObservedWindow:(UIWindow *)window
 {
+    if (!linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ApplicationStateTrackerDoesNotObserveWindow))
```
It would be ideal if we could detect the use of Aspects.framework and generate some kind of log message here. But that is likely to be hard/stupid/fragile, so may cause more trouble than its worth.

brentfulgham reviewed (approved):
This was what I was thinking about as a possible workaround, too. Looks good to me! r=me

jernoble committed:
REGRESSION (275943@main): Apps using Aspects framework can crash after ApplicationStateTracker adds KVObserver
https://bugs.webkit.org/show_bug.cgi?id=272532
rdar://125932015

Reviewed by Brent Fulgham.

Aspects.framework will crash if it tries to add a "hook" for KVO after an observer has already
been added to the target object. Add a linked-on-or-after check which disables this KVO, giving
apps time to move off of using Aspects in their production code.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(-[WKUIWindowSceneObserver setObservedWindow:]):

Canonical link: https://commits.webkit.org/277396@main
